Update distributed tracing docs (#23313)
dotnet/docs: docs/core/diagnostics/distributed-tracing-concepts.md

---
title: Distributed tracing concepts - .NET
description: .NET distributed tracing concepts
ms.date: 03/14/2021
---

# Distributed Tracing

Distributed tracing is a diagnostic technique that helps engineers localize failures and 
performance issues within applications, especially those that may be distributed across
multiple machines or processes. See the [Distributed Tracing Overview](distributed-tracing.md)
for general information about where distributed tracing is useful and example code to get
started.

## Concepts

### Traces and Activities

> Each time a new request is received by an application a new trace can be initiated. In

Reviewer: the application can join the existing trace (if there is one from the incoming request) or a new trace can be initiated

Author: I reworded to this to say it can be associated with a trace without explicitly saying the trace is new. I worry that the reader doesn't have enough context yet to understand what it means to join an existing trace or that trace information flows with the request.

occuring on that thread, accessible via 
<xref:System.Diagnostics.Activity.Current%2A?displayProperty=nameWithType>. The current activity
automatically flows along all synchronous calls on a thread as well as following async calls
that are processed on different threads. If Activity A is the current activity on a thread and
code starts a new Activity B then B becomes the new current activity on that thread. By default
activity B will also treat Activity A as its parent. When Activity B is later stopped activity
A will be restored as the current Activity on the thread. When an Activity is started it
captures the current time as the 
<xref:System.Diagnostics.Activity.StartTimeUtc%2A?displayProperty=nameWithType>. When it
stops <xref:System.Diagnostics.Activity.Duration%2A?displayProperty=nameWithType> is calculated
as the difference between the current time and the start time.

### Coordinating across process boundaries

In order to track work across process boundaries Activity parent IDs need to be transmitted across
the network so that the receiving process can create Activities that refer to them. When using

> W3C standard Activity IDs (<xref:System.Diagnostics.ActivityIdFormat.W3C%2A?displayProperty=nameWithType>)

Reviewer 2: ```suggestion
W3C standard Activity IDs (<xref:System.Diagnostics.ActivityIdFormat.W3C?displayProperty=nameWithType>)
```

> .NET will also use [W3C standard networking protocols](https://www.w3.org/TR/trace-context/) to

Reviewer: ```suggestion
.NET will also use [W3C TraceContext](https://www.w3.org/TR/trace-context/) to
```

transmit this information. When using the 

> <xref:System.Diagnostics.ActivityIdFormat.Hierarchical%2A?displayProperty=nameWithType> ID format

Reviewer 2: ```suggestion
<xref:System.Diagnostics.ActivityIdFormat.Hierarchical?displayProperty=nameWithType> ID format
```

.NET uses a custom request-id HTTP header to transmit the ID. Unlike many other language runtimes

> .NET in-box libraries natively understand how to decode and encode Activity IDs on HTTP messages

Reviewer 3: might be a good idea to clarify what is "in-box" libraries. You are referring to ASP.NET Core and HttpClient in .NET Core right?

later. There are several telemetry collection libraries that can do this task such as
[Application Insights](https://docs.microsoft.com/azure/azure-monitor/app/distributed-tracing),
[OpenTelemetry](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started/README.md),
or a library provided by a 3rd party telemetry or APM vendor. Alternately developers can author
their own custom Activity telemetry collection by using
<xref:System.Diagnostics.ActivityListener?displayProperty=nameWithType> or
<xref:System.Diagnostics.DiagnosticListener?displayProperty=nameWithType>. ActivityListener
supports observing any Activity regardless whether the developer has any a-priori knowledge about it.
This makes ActivityListener a simple and flexible general purpose solution. By contrast using
DiagnosticListener is a more complex scenario that requires the instrumented code to opt-in by
invoking <xref:System.Diagnostics.DiagnosticSource.StartActivity%2A?displayProperty=nameWithType> and
the collection library needs to know the exact naming information that the instrumented code

> used when starting it. Using DiagnosticSource and DiagnosticListener allows the creator

Reviewer 3: We could state that the in-box libraries publish DiagnosticSource events to pass things like the HttpContext instances, which may be leveraged by listeners to extract additional information to enrich the Activity.

Author: I think I will avoid going to into that level of detail here. I agree there is lots more we could say about DiagnosticSource which could probably could fill its own pages.

and listener to exchange arbitrary .NET objects and establish customized information passing
conventions.

### Sampling

For improved performance in high throughput applications, distributed tracing on .NET supports

> sampling only a subset of requests rather than recording all of them. For activites created with

Reviewer 3: ```suggestion
sampling only a subset of traces rather than recording all of them. For activites created with
```

the recommended <xref:System.Diagnostics.ActivitySource.StartActivity%2A?displayProperty=nameWithType>

> API, logging libraries can control sampling with the

Reviewer 3: ```suggestion
API, telemetry collection libraries can control sampling with the
```
prefer to to avoid "logging"

Author: arg, thanks, I missed this one : ) I had been trying to change it everywhere

<xref:System.Diagnostics.ActivityListener.Sample%2A?displayProperty=nameWithType> callback.
The logging library can elect not to create the Activity at all, to create it with minimal
information necessary to propagate distributing tracing IDs, or to populate it with complete
diagnostic information. These choices trade-off increasing performance overhead for
increasing diagnostic utility. Activities that are started using the older pattern of invoking

> <xref:System.Diagnostics.Activity..ctor%2A?displayProperty=nameWithType> and 

Reviewer 2: ```suggestion
<xref:System.Diagnostics.Activity.%23ctor%2A?displayProperty=nameWithType> and 
```
